### ISSUE TITLE
Improving K-9 Logging

### ISSUE THREAD

**Issue by philipwhiuk** [k9mail/k-9#1732]
### Expected behavior

Users should easily be able to report issues, with enough detail to allow us to debug them.

### Actual behavior

We get issues opened with little to no information. The recommended way to diagnose problems involves (for all recent Android devices) connecting to `adb`. This essentially restricts bug reports to Android developers only.

### Problems arising

In a normal app which is funded by a development team, you'd use something like ACRA which then pumps data to a back-end server.  K-9 doesn't have this option. It's an open source project, there's no hosting or funding available for that kind of solution.

Often the bugs we see are not crashes per see, which are fairly straightforward to solve, but instead Doze / data usage related.

### Possible solution 1 - Save logcat output

The reason we can't recommend aLogCat is because it no longer has access to our logs. Therefore we must read the logs ourselves. Obviously if we crash just reading them is insufficient. So we also have to write them to a file.

One option is to implement a rolling logging solution. This would:

* Run logcat on the device in our process. It turns out this is fairly straightforward, if rather ugly: 

http://www.journal.deviantdev.com/android-log-logcat-to-file-while-runtime/

* Periodically roll the log file. We don't want to take up too much space. This could be a configurable option - users with complex problems could increase the amount stored.

* Allow the user to save logs to external storage

Our issue form would then say:

    Upload the the latest log file from K-9's directory on your phone.

We need to carefully test this to make sure that the background logging itself doesn't cause any issues.

### Possible solution 2 - Stop using `android.util.Log`

The alternative is more drastic. We implement a new logging architecture. This would give us full control - we could flush to disk periodically (and when getting an uncaught exception) rather than all the time.

**Comment by cketti** [comment]
I'm for option 2 using [Timber](https://github.com/JakeWharton/timber).

In debug builds we could output to Android's system log. In release builds we write log messages to a file. I'd like us to display a warning message in the UI when file logging is enabled. That should reduce the chances of users leaving file logging enabled accidentally.

**Comment by jsone-studios** [comment]
I would suggest to use [crashlytics](https://fabric.io/kits/android/crashlytics). It is something like ACRA but you do not need to host your own backend und its completly free.

**Comment by GoneUp** [comment]
I think this can be closed, since Timber is included now in the lib and the main app.

**Comment by philipwhiuk** [comment]
Nope, we still have work to do. We need to start writing Timber logs to a file.

**Comment by philipwhiuk** [comment]
Before we release the next stable we should try and include file-based logging. There's some instructions here.

http://www.sureshjoshi.com/mobile/file-logging-in-android-with-timber/

I couldn't get it to work at the time. 

For debug mode we should use the current ADB strategy. For releases we should use a rolling log file. Ideally it should be in a folder they can get to with a file explorer so that they can provide it here. If necessary we can make the app share the file like an attachment download to do this.

**Comment by cketti** [comment]
I want us to show some kind of permanent hint when file logging is enabled. Having it enabled has the potential to slow down the device and dramatically increase wear on internal storage.

Maybe an ongoing notification?

**Comment by fobo66** [comment]
I would like to suggest [Lumberjack](https://github.com/MFlisar/Lumberjack) project that provides special Timber.Tree for file logging based on [logback](https://github.com/qos-ch/logback) and [slf4j](https://github.com/qos-ch/slf4j). We have possibility to flexibly configure amount of files, their size and name pattern (based on date or number).

I can help with integrating this in project.

**Comment by cketti** [comment]
@fobo66: Pull requests for this are very welcome. But please note that we won't make the feature available to users before the requirement to display a persistent notification is implemented. Another option might be to automatically disable logging to a file after a certain amount of time.

**Comment by 1000283** [comment]
Yes, **please** enable a freakin' logfile.
